v0/destinations/ga4/transform.js

```diff
@@ -0,0 +1,414 @@
+        break;
+    }
+  } else {
+    // custom events category
+    // Event names are case sensitive
+    if (isReservedWebCustomEventName(event)) {
+      throw new CustomError(
+        "[Google Analytics 4] track:: Reserved custom event names are not allowd",
+        400
+      );
+    }
+
+    if (isReservedWebCustomPrefixName(event)) {
+      throw new CustomError(
+        "[Google Analytics 4] track:: Reserved custom prefix names are not allowd",
+        400
+      );
+    }
+
+    payload.name = event;
+
+    // all extra parameters passed is incorporated inside params
+    let customParameters = {};
+    customParameters = extractCustomFields(
+      message,
+      customParameters,
+      ["properties"],
+      GA4_RESERVED_PARAMETER_EXCLUSION
+    );
+    if (!isEmptyObject(customParameters)) {
+      customParameters = flattenJson(customParameters);
+      payload.params = {
+        ...payload.params,
+        ...customParameters
+      };
+    }
+    removeReservedParameterPrefixNames(payload.params);
+  }
+
+  // take GA4 user properties
+  let userProperties = {};
+  userProperties = extractCustomFields(
+    message,
+    userProperties,
+    ["user_properties"],
+    GA4_RESERVED_USER_PROPERTY_EXCLUSION
+  );
+  if (!isEmptyObject(userProperties)) {
+    userProperties = flattenJson(userProperties);
+    rawPayload.user_properties = userProperties;
+  }
+
+  removeReservedUserPropertyPrefixNames(rawPayload.user_properties);
+
+  rawPayload = { ...rawPayload, events: [payload] };
+
+  // build response
+  const response = defaultRequestConfig();
+  response.method = defaultPostRequestConfig.requestMethod;
+  response.endpoint = ENDPOINT;
+  response.headers = {
+    HOST: "www.google-analytics.com",
+    "Content-Type": "application/json"
+  };
+  response.params = {
+    api_secret: Config.apiSecret
+  };
+  if (Config.measurementId) {
+    response.params.measurement_id = Config.measurementId;
+  } else {
+    response.params.firebase_app_id = Config.firebaseAppId;
+  }
+  response.body.JSON = rawPayload;
+  return response;
+};
+
+const process = async event => {
+  const { message, destination } = event;
+
+  if (!destination.Config.apiSecret) {
+    throw new CustomError("API Secret not found. Aborting ", 400);
+  }
+  if (!destination.Config.measurementId && !destination.Config.firebaseAppId) {
+    throw new CustomError(
+      "measurementId or firebaseAppId must be provided. Aborting",
+      400
+    );
+  }
+
+  if (!message.type) {
+    throw new CustomError(
+      "Message Type is not present. Aborting message.",
+      400
+    );
+  }
+
+  const messageType = message.type.toLowerCase();
+  let response;
+  switch (messageType) {
+    case EventType.TRACK:
+      response = await trackResponseBuilder(message, destination);
+      break;
+    default:
+      throw new CustomError(`Message type ${messageType} not supported`, 400);
+  }
+  return response;
+};
+
+module.exports = { process };
```

Comment:
Can't we support router transformation?

Reply:
we are planning to use CDK for GA4 later due to which we decided not to do routerTransform for now.